mino/minows/mod.go

```diff
@@ -0,0 +1,125 @@
+package minows
+
+import (
+	"github.com/rs/zerolog"
+	"go.dedis.ch/dela"
+	"go.dedis.ch/dela/serde/json"
+	"regexp"
+	"strings"
+
+	"github.com/libp2p/go-libp2p"
+	"github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/protocol"
+	ma "github.com/multiformats/go-multiaddr"
+	"go.dedis.ch/dela/mino"
+	"go.dedis.ch/dela/serde"
+	"golang.org/x/xerrors"
+)
+
+// Minows
+// - implements mino.Mino
+type minows struct {
+	logger zerolog.Logger
+
+	myAddr   address // publicly reachable address
+	host     host.Host
+	context  serde.Context
+	segments []string       // namespace
+	rpcs     map[string]any // names
+}
+
+// newMinows creates a new Minows instance.
+// listen: local listening address in multiaddress format,
+// e.g. /ip4/0.0.0.0/tcp/80/ws
+// public: public dial-able address in multiaddress format,
+// e.g. /dns4/p2p-1.c4dt.dela.org/tcp/443/wss
+// secret: private key representing this mino instance's identity
+func newMinows(listen, public ma.Multiaddr, secret crypto.PrivKey) (*minows,
+	error) {
+	// create publicly reachable address
+	id, err := peer.IDFromPrivateKey(secret)
+	if err != nil {
+		return nil, xerrors.Errorf("could not derive identity: %w", err)
+	}
+	myAddr, err := newAddress(public, id)
+	if err != nil {
+		return nil, xerrors.Errorf("could not create address: %v", err)
+	}
+	// start listening
+	h, err := libp2p.New(libp2p.ListenAddrs(listen), libp2p.Identity(secret))
+	if err != nil {
+		return nil, xerrors.Errorf("could not create host: %v", err)
+	}
+	return &minows{
+		logger:   dela.Logger.With().Str("mino", myAddr.String()).Logger(),
+		myAddr:   myAddr,
+		segments: nil,
+		host:     h,
+		context:  json.NewContext(),
+		rpcs:     make(map[string]any),
+	}, nil
+}
+
+func (m *minows) GetAddressFactory() mino.AddressFactory {
+	return addressFactory{}
+}
+
+func (m *minows) GetAddress() mino.Address {
+	return m.myAddr
+}
+
+func (m *minows) WithSegment(segment string) mino.Mino {
+	if segment == "" {
+		return m
+	}
+
+	return &minows{
+		myAddr:   m.myAddr,
+		segments: append(m.segments, segment),
+		host:     m.host,
+		rpcs:     make(map[string]any), // don't copy existing rpcs
+		context:  m.context,
+	}
+}
+
+func (m *minows) CreateRPC(name string, h mino.Handler, f serde.Factory) (mino.RPC, error) {
+	pattern := regexp.MustCompile("^[a-zA-Z0-9]+$")
+	// validate namespace once
+	if len(m.rpcs) == 0 {
```

Comment:
I wonder whether we should put that check in the `minows.WithSegment` method. The only downside is that `WithSegment` doesn't return an error, and you'd have to panic.

But having an invalid segment name would be a valid reason for me to panic.